### ISSUE TITLE
1 second delay in partial terminal window update using VGL.

### ISSUE THREAD

**Issue by marioroy** [TurboVNC/turbovnc#364]
Thank you for TurboVNC with VirtualGL capabilities.

Unfortunately, enabling useVGL causes issues with Gnome Terminal. The last line isn't updated until 1 second later. Sometimes, the cursor is partially drawn and completed drawing 1 second later. The worst is the one second lag (randomly) for keyboard input, from the time the keyboard is pressed to the character fully displayed inside the terminal.

Server:
TurboVNC 3.0.3
NVIDIA Driver 530.30.02
NVIDIA Graphics RTX 3070
Clear Linux OS (Release 38630)
GNOME Version 43.3
Windowing System X11

Client:
macOS Big Sur 11.7.3 (Intel)
TurboVNC 3.0.3

Client:
The same behavior occurs running vncviewer on Clear Linux.

I filed a [bug ticket](https://github.com/clearlinux/distribution/issues/2859) at Clear Linux due to `xkbcomp` failing. Otherwise, TurboVNC works great after making the 3 symbolic links.

I will try another NVIDIA driver release (starting with 525, 520, 510).

_The rest is off-topic, in the event someone wants to try TurboVNC on Clear Linux._

There used to be regular live desktop image ISOs. See the steps [here](https://github.com/clearlinux/distribution/issues/2830) for making a functional live desktop USB.

Using NVIDIA graphics? See [automation](https://github.com/marioroy/nvidia-driver-on-clear-linux) for installing the NVIDIA driver. This requires booting the live USB into text mode by pressing the letter `e` at the boot menu. Prepend "3 " with a space to the list of kernel arguments. Instructions are provided on the screen for running the installer. Install the OS to a different USB drive.

I installed TurboVNC using the RPM path. You will need the `package-utils` bundle. There are issues with the RPM embedded script not supporting Clear Linux. So, the `--noscripts` option is specified.

```bash
sudo swupd bundle-add package-utils x11-server x11-tools
sudo rpm -ivh --noscripts --nodeps turbovnc-3.0.3.x86_64.rpm

# To uninstall
sudo rpm -e --noscripts turbovnc
```

I installed `VirtualGL-3.1`, obtained from the [web](https://sourceforge.net/projects/virtualgl/files/).

```bash
sudo rpm -ivh --nodeps VirtualGL-3.1.x86_64.rpm

# To uninstall
sudo rpm -e VirtualGL
```

**Comment by marioroy** [comment]
The bottom of the terminal window is not refreshed after exiting `man ls`.

<img width="1017" alt="image" src="https://user-images.githubusercontent.com/7031085/227425071-136149d7-d76e-4a1b-9a56-d134e837e5d7.png">

**Comment by marioroy** [comment]
There is a one second delay before the last line (Manual page ls...) is drawn.

<img width="999" alt="image" src="https://user-images.githubusercontent.com/7031085/227425496-f46fad23-d3b6-4c82-846c-51cc59b6170e.png">

**Comment by marioroy** [comment]
The one-second delay (randomly) for screen updates is awkward, using VGL. I tested NVIDIA driver 510.08.03, 525.89.02, 530.30.02, and 530.41.03.

Non-VGL works fantastic.

**Comment by marioroy** [comment]
[Wandering Pixel](https://github.com/justinrdonnelly/wandering-pixel) does not work in GNOME 43.3, something I wanted to try. The issue goes away if I run glxgears in the background. I changed the TVNC_VGLRUN line in xstartup.turbovnc to the following, to limit the fps.

```text
TVNC_VGLRUN="vglrun +wm +sync -sp -fps 60 -np 2
```

**Comment by dcommander** [comment]
Another user reported a different but possibly related issue (#362) with GNOME Terminal, VGL, and TurboVNC, but I was unable to reproduce it, and it went away with the latest nVidia driver. In attempting to reproduce that issue, I did not observe this issue either. Since Clear Linux isn't [officially supported](https://turbovnc.org/Documentation/OSSupport) by TurboVNC, it is incumbent upon the user community to diagnose the issue. There is very little I can do about it unless the issue is reproducible on a supported platform. I test about a dozen different Linux distributions in my lab using VMWare, including a couple (openSUSE Leap and Fedora) that aren't officially supported, but I can only test the nVidia drivers on CentOS and Rocky.

**Comment by marioroy** [comment]
Next, I tried Fedora 37. Here, the problem is worst running VGL.

turbovnc-3.0.3-20230227.x86_64
VirtualGL-3.1-20230315.x86_64

![image](https://user-images.githubusercontent.com/7031085/227641502-d5bc7cd6-0e67-4ea0-9a71-550f7f876ba7.png)

I clicked on the Terminal icon, but no terminal is displayed.

![image](https://user-images.githubusercontent.com/7031085/227642322-d8b1151d-04a0-4795-bb3a-c2bb394abe80.png)

**Comment by marioroy** [comment]
I tried Ubuntu 20.04.6 LTS (Focal Fossa).

1. turbovnc_3.0.3_amd64.deb
2. virtualgl_3.1_amd64.deb
3. NVIDIA driver 515.86.01
4. GNOME version 3.36.8
5. Windowing system X11

![image](https://user-images.githubusercontent.com/7031085/227711008-dc282b42-a2d8-4dc7-b3ce-0e07b5350998.png)

I moved the window around, which refreshed the TurboVNC window. I tried again: man ls. Notice the line "do not ignore entries starting with" not completely drawn.

![image](https://user-images.githubusercontent.com/7031085/227711334-1d603e24-b035-4ffe-9108-81428584642f.png)

I exited the man page. The last line is not cleared.

![image](https://user-images.githubusercontent.com/7031085/227711559-f423a03e-c432-466a-8db2-9111c5b42518.png)

**Comment by dcommander** [comment]
What do you mean "all is fine"? You mean that you no longer see the issue, or you don't see the issue with that specific viewer?

**Comment by marioroy** [comment]
For Ubuntu, I tested initially using `vncviewer` on the same machine where running `vncserver -vgl` or `vncserver`. I connected to 127.0.0.1:1.

I later ran the TurboVNC / Viewer 3.0.3 application on the Mac (connecting remotely to Ubuntu) and gnome-terminal ran fine, no issues using VirtualGL. Today, I will try the NVIDIA 515.86.01 driver on Clear Linux.

**Comment by dcommander** [comment]
OK, now you seem to be describing multiple issues. I need to understand exactly which symptoms are observed with which configurations. Please specify the client/server configurations with which you observe:
1. The issue whereby GNOME Terminal does not appear at all when run with VirtualGL
2. The issue whereby characters typed into GNOME Terminal (when run with VGL) appear only after a delay 

I have not observed either issue, so you're going to have to help me figure out exactly which component is failing on which platform.

Also, if you are using any non-default TurboVNC settings, then please disclose them. Also please tell me the approximate latency and bandwidth of your network.

**Comment by marioroy** [comment]
For all three OS platforms tested, I installed TurboVNC 3.0.3 and VirtualGL 3.1 with default settings. The network is Wi-Fi on the Mac. No issues with network bandwidth (~ 54MB). The server host is connected via wire using 1 Gigabit interface to the same router.

1. The issue whereby GNOME Terminal does not appear is on Fedora Linux 37. I tested vncviewer on the same box (locally) to 127.0.0.1:2 including on the Mac connecting remotely to 192.168.1.34:2. VGL appears broken, unusable.

2. The issue whereby characters typed into GNOME Terminal (when run with VGL) appearing after 1 second delay occurs on Clear Linux (yes, I know not officially supported) running vncviewer locally to 127.0.0.1:1 and from the Mac connecting remotely to 192.168.1.34:1. The issue also occurs on Ubuntu 20.04.6 running vncviewer locally on the same box connecting to 127.0.0.1:1. Locally, running on the same box where vncserver is running. 

The only time I have experienced success with VGL is running VNC viewer on the Mac connecting remotely 192.168.1.34:1 to Ubuntu 20.04.6.

There's a warning running vncserver on Fedora 37 stating unable to connect to :1, already used by X11 (some .X11-unix socket file exist in /tmp). So, it automatically selects :2. To get VGL to work, I needed to pass the -d :2 option to vglrun inside the xstartup script. 

In all 3 OS tested { Clear Linux, Fedora Linux 37, and Ubuntu 20.04.6 }, non-VGL works great. No issues at all and simply amazing. BTW, I connected directly, no tunneling during testing. The SSH tunneling is something I tried subsequently getting audio to work. So amazing really. Youtube video is smooth and no hiccups to audio. Even holding down the backspace in GNOME terminal emits beeps until releasing the backspace. There were no gaps or hiccups to audio.

I have never seen anything so fast and smooth until trying TurboVNC. Great work.

I also tried OpenCL and CUDA displaying hundreds of Mandelbrot images automatically. It works with/non VGL which is great. I hadn't realize that VGL isn't needed to run OpenCL/CUDA on the GPU. I also ran glxspears64 VGL/non-VGL mainly to witness the effect of running VGL.

I'm hoping that all is clear now, nothing left out. All defaults, except on Fedora 37 needing -d :2 to vglrun.

**Comment by dcommander** [comment]
OK, but that doesn't solve the issue, so I'm not sure why you closed it.

**Comment by marioroy** [comment]
I felt bad for any time lost, especially when Clear Linux and Fedora 37 are not in official supported list. Though, the issue may creep up in the next major supported list distribution.

**Comment by marioroy** [comment]
Sir, thank you. I miss-read the vglrun help usage. Sure enough, defaulting to :0.0 is the clue along with mentioning DRI. I recalled momentarily puzzled at the time, because wrongly assuming the number matching vncserver.

```text
-d <d>    : <d> = the X display/screen or DRI device to use for 3D rendering
           [default = :0.0]
```

On Fedora 37, the (-d :1) option is needed as vglrun will not run without it. The reason is that /tmp/.X11-unix/X1 exists and running vncserver  emits a warning. Ditto without the (-vgl) option.

```text
$ ./vncserver -vgl

WARNING: fedora-pc:1 is taken because of /tmp/.X11-unix/X1
Remove this file if there is no X server fedora-pc:1

Desktop 'TurboVNC: fedora-pc:2 (mario)' started on display fedora-pc:2
```

It seems that vglrun does not know to use (-d :1) on its own without manual intervention. I removed the (-d :1) option to capture the log. The session exits immediately.

```text
xstartup.turbovnc: Executing vglrun +wm /etc/X11/xinit/Xsession "/usr/bin/gnome-session"
Killing Xvnc process ID 6463
```

Fedora 37 (vncserver -vgl) exhibits the same 1 second delay issue/behaviour seen on Clear Linux. The workaround, changing the height of the terminal window to 80x25 resolves the issue using the vncviewer on the mac (remote connection to the Fedora box).

```bash
TVNC_VGLRUN="vglrun +wm -d :1"
```

Note: The workaround 80x25 using vncviewer on Fedora (same box), connecting locally to 127.0.0.1:2, does not work (1 second delay or unfinished refresh).
